### PR TITLE
Add deepCopy to GATKRead.

### PR DESCRIPTION
**Pull request by cmnbroad**

### CODE REVIEW

File: src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java [hunk 1]
```diff
@@ -1394,4 +1394,59 @@ private int parseSAMStringFlags(String samString) {
         Assert.assertTrue(m.find());
         return Integer.parseInt(m.group(2)); // use the second capture group in the reg ex above since we don't want the tab
     }
+
+    @DataProvider(name="copyData")
+    public Object[][] getCopyData() {
+        List<Object[]> testCases = new ArrayList<>();
+
+        testCases.add(new Object[]{basicReadBackedBySam()});
+        testCases.add(new Object[]{basicReadBackedByGoogle()});
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java]: add one for ADAM reads `basicReadBackedByADAMRecord`

File: src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java [hunk 2]
```diff
@@ -1394,4 +1394,59 @@ private int parseSAMStringFlags(String samString) {
...
+        // reverseStrand is nested as LinearAlignment->Position->reverseStrand
+        read.setIsReverseStrand(false);
+        deepCopy = read.deepCopy();
+        Assert.assertTrue(read.isReverseStrand() == deepCopy.isReverseStrand());
+        read.setIsReverseStrand(true);
+        Assert.assertFalse(read.isReverseStrand() == deepCopy.isReverseStrand());
+        Assert.assertFalse(read.equals(deepCopy));
+
+        // attributes
+        deepCopy = read.deepCopy();
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java]: these reassignments suggest splitting this test method into 3 methods `testDeepCopyStrand`, `testDeepCopyAttributes` and `testDeepCopyPosition`

Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java]: but it's optional - it's testing code so i dont feel strongly. your call

File: src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java [hunk 3]
```diff
@@ -1394,4 +1394,59 @@ private int parseSAMStringFlags(String samString) {
...
+        return testCases.toArray(new Object[][]{});
+    }
+
+    @Test(dataProvider="copyData")
+    public void testShallowCopy(final GATKRead read) {
+        Assert.assertTrue(read.equals(read.copy()));
+
+        read.setIsReverseStrand(false);
+        GATKRead shallowCopy = read.copy();
+        Assert.assertTrue(read.isReverseStrand() == shallowCopy.isReverseStrand());
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java]: avoid assertTrue for comparisons, it's less informative than assertEquals

File: src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java [hunk 4]
```diff
@@ -1394,4 +1394,59 @@ private int parseSAMStringFlags(String samString) {
...
+        read.setIsReverseStrand(true);
+        Assert.assertFalse(read.isReverseStrand() == shallowCopy.isReverseStrand());
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java]: use `assertNotEquals`

**Comment by akiezun** [comment]
looks good. minor comments. Merge when done (ie rebase, squash, merge when done and test pass, of course)

File: src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java (hunk 1 above)
Comment (cmnbroad) [inline src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java]: That pulls on a long thread, and is the reason I asked yesterday about the status of ADAM backed reads. It currently doesn't have copy or equals implementations, so copy always results in a SAMRecord backed read; the records are never equal to themselves; it never updates its backing AlignmentRecord. I started down the path of fixing those and ran into problems with the constructors and converters. So I think these things should be part of #1189, but let me know if you want it to be part of this.

Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java]: fair enough - please add a note to #1189 to add a test here

Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java]: nevermind. i just did

**Pull request merged**